### PR TITLE
chore(devenv): introduced parameterized demo HTML

### PR DESCRIPTION
**Pull request by asudoh**
## Overview

This PR switches component markup code from HTML to ~~Dust~~ Handlebars, allowing to generate HTML snippets for multiple variants easier, if differences in those variants are small.

~~TODO:~~

- [x] ~~Try to align HTML file names to where they use to be, especially for major components that some projects may be using directly~~
- [x] ~~Update tests to refer to the built version of the HTML snippets~~

### Added

* Support for [`component.config.js`](https://fractal.build/guide/core-concepts/configuration-files#javascript-module-format), handling a subset of Fractal properties as follows:
  * [`name`](https://fractal.build/guide/components/configuration#name)
  * [`label`](https://fractal.build/guide/components/configuration#label)
  * [`notes`](https://fractal.build/guide/components/configuration#notes)
  * [`context`](https://fractal.build/guide/components/configuration#context)
  * [`preview`](https://fractal.build/guide/components/configuration#preview) (Can be left `undefined`, if specified, should refer to a ~~Dust~~ Handlebars template file name in `demo/views`)

### Changed

* Component sample markup from HTML to ~~Dust~~ Handlebars
* The HTML snippet view in dev env, now it sends an AJAX request to fetch the snippet so the server code can (lazily) render the ~~Dust~~ Handlebars template
* The build tool to generate component HTML snippets, putting ~~Dust~~ Handlebars rendering logic in it

## Testing / Reviewing

Starting with button for the review will be the easiest, to see how things work.

Testing should make sure dev env is not broken for any components, and also make sure HTML build is not broken.

### Current goal

- [x] Make it easy to create demo HTML snippets for multiple variants that changes only small fraction of markup (“One template supporting multiple variants” scenario)

### Future goal (With separate PRs)

- Support SSR for several components where doing so is straightforward
- Template instantiation with design tokens
- Introduce client-side template, as a separate set of templates, once [the spec](https://github.com/w3c/webcomponents/blob/gh-pages/proposals/Template-Instantiation.md) is standardized (Expecting very future)

### Non-goal

- Introducing a way to auto-convert the templates to different AJAX frameworks (Angular/React) - Not realistic

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -26,7 +25,7 @@ app.use(webpackHotMiddleware(compiler));
 
 const fractal = Fractal.create();
 fractal.components.set('path', path.join(__dirname, 'src/components'));
-fractal.components.set('ext', '.html');
+fractal.components.set('ext', '.dust');
```
Comment (joshblack) [inline server.js]: Is there any way we could also support `hbs` as well?

**Comment by joshblack** [comment]
Going through this now, looks great 😄 

Was curious about two things, namely:

- Is there a way to support handlebars here so Watson Health can contribute?
- How would we want someone to consume either template?

File: server.js (hunk 1 above)
Comment (scottnath) [reply server.js]: fractal only allows on templating language

**Comment by asudoh** [comment]
@joshblack [Yes for the former](https://github.com/carbon-design-system/carbon-components/pull/579/commits/6dabc3badb46ea142432e6f1c05cf721e67d9ecb). As of now the template is built into HTML as the dist files. Direct consumption of templates is not covered in this PR as this PR focuses on "supporting multiple variants per template" scenario, but definitely in consideration - Along with design token stuff which I think we talked elsewhere in GH (but I cannot rattle off the link).

File: src/components/button/button.hbs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<button class="bx--btn bx--btn--{{variant}}{{#if small}} bx--btn--sm{{/if}}" type="button">Button</button>
```
Comment (joshblack) [inline src/components/button/button.hbs]: Could the variants (like small/big) just be separate hbs files?

I think Watson Health does `component--variant.hbs` as the pattern

Comment (asudoh) [reply src/components/button/button.hbs]: @joshblack Yes, that’s seen in tables in this PR.

Comment (asudoh) [reply src/components/button/button.hbs]: BTW given you brought up separate files for variants - Wondering what you imagined using templates for? Text node values? Attributes? Repeat?

**Comment by joshblack** [comment]
Was going through some of this now, noticed:

1. The demo-container seems to add a lot of height to each example that causes the overflow to exist even though it is redundant in some cases

screen shot 2018-02-21 at 10 58 23 am

screen shot 2018-02-21 at 10 58 28 am

2. Reloading the page seems to stall, might just be my machine though

3. I don't seem to be able to examine styles on the page, is there something I'm doing wrong on my end?

screen shot 2018-02-21 at 10 59 50 am

**Review by marijohannessen — COMMENTED** [review]
Would this get rid of the `html` files completely? Just worried about the users that are depending on them as of now - the website is also using the `html` files directly to show each component (although we are going to change this by v9). Could we surface both? If not, we'll have to have this in a separate v9 branch until the website is updated. Also not having the variations in separate files will potentially make it hard to show the different variations the way we do it right now on the website.

**Comment by marijohannessen** [comment]
@joshblack I'm experiencing the same issues as you are

**Comment by joshblack** [comment]
Totally agreed that there should be some notion of the `html` files for vanilla, how are we pulling the examples in now for the site? Worst case we can have a build step that adds the html files for all the variants since that would mean just compiling the templates with the config files

With respect to the different variants, I think Watson Health is doing:

```
# Actual component template that a server could use
component.hbs

# All component variants
component--all.hbs

# Specific component variant
component--variant.hbs
```

So I think we could easily follow that model to allow multiple variants in one template file/demo.

**Comment by marijohannessen** [comment]
Right now we're just looking for the specific HTML files in `carbon-components/src/components` and rendering the content of the file in the component example. We could probably easily switch it to pull the `hbs` files instead but we would still need a differentiation between the variations.

**Comment by joshblack** [comment]
I'm not sure what the idea is for the site looking forward, and this is totally outside the scope of this issue, but could be a good opportunity for a "to-be" scenario that introduces tabs to the top of the code snippets that let's you select a variant for how to use the component and copy the code for your use-case.

e.g., tabs like HTML - Handlebars - React - Angular that someone could click and copy+paste.

**Comment by marijohannessen** [comment]
Yup, that's part of the plan for V9 (issue here https://github.ibm.com/Bluemix/carbon-issues/issues/451) 🙂  I just want to make sure that the website gets updated accordingly to any updates that happen here - removing the HTML files completely feels like a pretty drastic step.

**Comment by joshblack** [comment]
That's so exciting!

**Comment by asudoh** [comment]
Thanks @joshblack and @marijohannessen for jump in and discussing this thing! Wrt the HTML thing - `html:source` Gulp task has been changed so that it generates almost the same set of HTML as we used to have. Also the newly added `/code/:component` Express route [calls](https://github.com/asudoh/carbon-components/blob/6dabc3badb46ea142432e6f1c05cf721e67d9ecb/server.js#L119-L120) another newly added [(currently) internal `template.render()` tooling API](https://github.com/asudoh/carbon-components/blob/6dabc3badb46ea142432e6f1c05cf721e67d9ecb/tools/templates.js#L71-L108) to generate the HTML snippet in the code example UI in the dev env.

Hope this clarifies the question regarding HTML. Please don’t hesitate to ping me either of you have any further questions/concerns. Thanks!

P.S. Will take a look at the issue of viewport height.

**Comment by asudoh** [comment]
Addressed the viewport height concern: https://github.com/carbon-design-system/carbon-components/pull/603

**Comment by asudoh** [comment]
From the feedback from our devs, changed the target from `master` to `v9-dev` to address concerns on minor breaking change wrt change in HTML file name. @joshblack Please don’t hesitate to speak up if you have any concerns and I can change the branch (back or another). Thanks!

**Comment by shinytoyrobots** [comment]
Review for v9

**Review by marijohannessen — APPROVED** [review]

**Review by tw15egan — APPROVED** [review]
This looks awesome! Nice work

**Review by joshblack — APPROVED** [review]

**Comment by asudoh** [comment]
Thanks everybody for reviewing!!

**Pull request merged**